lib/ts/recipe/webauthn/api/implementation.ts

```diff
@@ -0,0 +1,1030 @@
+                attestation: response.attestation,
+                pubKeyCredParams: response.pubKeyCredParams,
+                excludeCredentials: response.excludeCredentials,
+                rp: response.rp,
+                user: response.user,
+                authenticatorSelection: response.authenticatorSelection,
+            };
+        },
+
+        signInOptionsPOST: async function ({ tenantId, options, userContext }) {
+            const relyingPartyId = await options.config.getRelyingPartyId({
+                tenantId,
+                request: options.req,
+                userContext,
+            });
+            const relyingPartyName = await options.config.getRelyingPartyName({
+                tenantId,
+                userContext,
+            });
+
+            // use this to get the full url instead of only the domain url
+            const origin = await options.config.getOrigin({
+                tenantId,
+                request: options.req,
+                userContext,
+            });
+
+            const timeout = DEFAULT_SIGNIN_OPTIONS_TIMEOUT;
+            const userVerification = DEFAULT_SIGNIN_OPTIONS_USER_VERIFICATION;
+            const userPresence = DEFAULT_SIGNIN_OPTIONS_USER_PRESENCE;
+
+            const response = await options.recipeImplementation.signInOptions({
+                userVerification,
+                userPresence,
+                origin,
+                relyingPartyId,
+                relyingPartyName,
+                timeout,
+                tenantId,
+                userContext,
+            });
+
+            if (response.status !== "OK") {
+                return response;
+            }
+
+            return {
+                status: "OK",
+                webauthnGeneratedOptionsId: response.webauthnGeneratedOptionsId,
+                rpId: relyingPartyId,
+                createdAt: response.createdAt,
+                expiresAt: response.expiresAt,
+                challenge: response.challenge,
+                timeout: response.timeout,
+                userVerification: response.userVerification,
+            };
+        },
+
+        signUpPOST: async function ({
+            webauthnGeneratedOptionsId,
+            credential,
+            tenantId,
+            session,
+            shouldTryLinkingWithSessionUser,
+            options,
+            userContext,
+        }) {
+            const errorCodeMap = {
+                SIGN_UP_NOT_ALLOWED:
+                    "Cannot sign up due to security reasons. Please try logging in, use a different login method or contact support. (ERR_CODE_007)",
+                INVALID_AUTHENTICATOR_ERROR:
+                    "The device used for authentication is not supported. Please use a different device. (ERR_CODE_26)",
+                INVALID_CREDENTIALS_ERROR:
+                    "The credentials are incorrect. Please make sure you are using the correct credentials. (ERR_CODE_25)",
```

Comment:
This is strange, is this required? The nvalid_X errors map to `SIGN_UP_NOT_ALLOWED` with a reason?